src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -187,6 +189,11 @@ private unsafe SocketError ProcessIOCPResult(bool success, int bytesTransferred,
 
             // Socket handle is going to post a completion to the completion port (may have done so already).
             // Return pending and we will continue in the completion port callback.
+            if (_singleBufferHandleState == SingleBufferHandleState.InProcess)
+            {
+                RegisterToCancelPendingIO(overlapped, cancellationToken); // must happen before we change state to Set to avoid race conditions
+                _singleBufferHandleState = SingleBufferHandleState.Set;
```

Comment:
Probably not an issue, but that this will lead to `_singleBufferHandle.Dispose()` in `Complete`.

Also note that after these changes, the only semantical difference between `ProcessIOCPResult` and `ProcessIOCPResultWithSingleBufferHandle` is the pinning of the handle. Isn't it worth to dedupe that code at this point?

Reply:
Yeah, it is. I am trying to clean this up, but it's a little tricky at points...